### PR TITLE
plugin/metadata: finish documentation

### PR DESCRIPTION
**Pull request by miekg**
Finish the README.md, add corner case in the IsLabel test and reword
some code comments slightly.

Generate the man-pages and add man/coredns-metadata.7 as well.

Signed-off-by: Miek Gieben <miek@miek.nl>

### CODE REVIEW

File: man/coredns-metadata.7 [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+By enabling \fImetadata\fR any plugin that implements metadata\.Provider interface \fIhttps://godoc\.org/github\.com/coredns/coredns/plugin/metadata#Provider\fR will be called for each DNS query, at beginning of the process for that query, in order to add it\'s own meta data to context\.
+.
+.P
+The meta data collected will be available for all plugins, via the Context parameter provided in the ServeDNS function\. The package (code) documentation has examples on how to inspect and retrieve metadata a plugin might be interested in\.
+.
+.P
+The meta data is added by setting a label with a value in the context\. These labels should be named \fBplugin/NAME\fR, where \fBNAME\fR is something descriptive\. The only hard requirement the \fImetadata\fR plugin enforces is that the labels contains a slash\. See the documentation for \fBmetadata\.SetValueFunc\fR\.
+.
+.P
+The value stored is a string\. The empty string signals "no meta data"\. See the documentation for \fBmetadata\.ValueFunc\fR on how to retrieve this\.
```
Comment (fturib) [inline man/coredns-metadata.7:19]: this info: `empty string signals "no meta data"`
we have 3 cases:
- Provider advertise a Label and when retrieving the value, that value is not empty (expected regular case)
- Provider advertise a Label and when retrieving the value, that value is empty
- Provider do not advertise a Label 

Which one do we call "no meta data", view from the consumer ?

Consumer knows if Label is advertised with ValueFunc not NIL
if consumer can retrieve a value for a Label with ValueFunc, then there is a meta data .. and maybe that value is empty.

IMO,  the "no meta data" case is when the Label is not advertised (ValueFunc return a nil)

File: man/coredns-rewrite.7 [hunk 2]
```diff
@@ -306,12 +306,16 @@ rewrites the first local option with code 0xffee, setting the data to "abcd"\. E
...
 .IP "" 0
 .
-.TP
-A variable data is specified with a pair of curly brackets \fB{}\fR\. Following are the supported variables
-{qname}, {qtype}, {client_ip}, {client_port}, {protocol}, {server_ip}, {server_port}\.
+.IP "\(bu" 4
+A variable data is specified with a pair of curly brackets \fB{}\fR\. Following are the supported variables: {qname}, {qtype}, {client_ip}, {client_port}, {protocol}, {server_ip}, {server_port}\.
+.
+.IP "\(bu" 4
+If the metadata plugin is enabled, then labels are supported as variables if they are presented within curly brackets\. the variable data will be filled with the value associated with that label\. If that label is not provided, the variable will be silently substitute by an empty string\.
```
Comment (fturib) [inline man/coredns-rewrite.7:313]: If that label is not provided, the rewrite rule will be ignored\.

File: plugin/metadata/README.md [hunk 3]
```diff
@@ -8,15 +8,20 @@
...
+metadata a plugin might be interested in.
 
-TODO: write about naming of the keys (labels).
-TODO: write about extracting and using
+The meta data is added by setting a label with a value in the context. These labels should be named
+`plugin/NAME`, where **NAME** is something descriptive. The only hard requirement the *metadata*
+plugin enforces is that the labels contains a slash. See the documentation for
+`metadata.SetValueFunc`. 
+
+The value stored is a string. The empty string signals "no meta data". See the documentation for
```
Comment (fturib) [inline plugin/metadata/README.md:23]: same note as above. I would just not mention "The empty string signals "no meta data". ".

File: plugin/metadata/metadata_test.go [hunk 4]
```diff
@@ -68,20 +68,23 @@ func TestLabelFormat(t *testing.T) {
...
 		{"/", false},
+		{"//", false},
 	}
 
 	for _, test := range labels {
-		if IsLabel(test.label) != test.isValid {
-			t.Errorf("Label %v is expected to have this validaty : %v - and has the opposite", test.label, test.isValid)
+		if x := IsLabel(test.label); x != test.isValid {
+			t.Errorf("Label %v expected %v, got: %v", test.label, test.isValid, x)
 		}
```
Comment (fturib) [inline plugin/metadata/metadata_test.go:88]: thanks for cleanup.

**Review by fturib — COMMENTED** [review]
I tried to explain what I think about the "empty" and "no metadata".

NOTE: for Rewrite, current coding will do:

- if Label is not advertised : just keep the rewrite (no edns0)
-if the Label is advertised and ValueFunc return empty (""), then the rewrite will happen and an edns0 entry will be created for the provide key and the empty "" value.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/metada..." ]
>NOTE: for Rewrite, current coding will do:
>
>- if Label is not advertised : just keep the rewrite (no edns0)
>-if the Label is advertised and ValueFunc return empty (""), then the rewrite will happen and an edns0 entry will be created for the provide key and the empty "" value.

I don't think that should be the contract. If you can't set metadata, i.e. your 
value is empty, you don't set anything. I can't think of a scenario where this 
would not be the case.

**Comment by fturib** [comment]
Hum.. I am not really convince it is better to merge Unknown and Empty. We just remove fine grain in the metadata information.
But I will not argue for long time.
So let's go like that.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/metada..." ]
>So let's go like that.

/merge
